### PR TITLE
Fixes #21392: Web-service endpoint is not available for the deployed EJB application

### LINKED ISSUE #21392: Web-service endpoint is not available for the deployed EJB application
A colleague of mine originally reported this bug for Glassfish 4.0 here:

[https://java.net/jira/browse/GLASSFISH-20661](https://java.net/jira/browse/GLASSFISH-20661)

Although I've recently updated that issue, it's doubtful that it will ever be looked at again, based on the activity of that issue's assignee(s), and the fact that it's marked as affecting version 4.0.

This bug still exists in Glassfish 4.1, so I'm creating a new issue for GF 4.1 here...

I have clarified the problem below:

Scenario:

JAR file "ws_ejb_ok.jar" contains a Stateless session bean implementation, annotated by @Stateless, that exposes a web-service endpoint by using a @WebService annotation.
The "name" attribute of the @Stateless annotation (value: "HelloEJBa") matches the "<ejb-name>" specified in the ejb-jar.xml descriptor.
Note that the <ejb-class> specified in the ejb-jar.xml descriptor matches the stateless session-bean implementation class.
Result: the JAR file deploys OK, the session bean and web-service endpoint are available and working. There is a message in the server.log file "EJB Endpoint deployed ws_ejb_ok listening at address at [http://](http://)...", confirming that the endpoint is available.

JAR file "ws_ejb_ng.jar" is almost exactly the same as JAR file "ws_ejb_ok.jar", except that the value of the "name" attribute of the @Stateless annotation (value "HelloEJBa") DOES NOT MATCH the "<ejb-name>" specified in the ejb-jar.xml descriptor (value "HelloEJBb").
Result: the JAR file deploys OK (and no errors in the server.log), and two stateless session-beans are deployed (which I believe is correct because they have different names), BUT the web-service endpoint IS NOT AVAILABLE.
The expected message "EJB Endpoint deployed ws_ejb_ng listening at address at [http://](http://)..." DOES NOT APPEAR in the server.log file.
Effectively the web-service is disabled by defining another session bean in the ejb-jar.xml descriptor that has the same class but different ejb-name.

I have fully explained the bug and it's cause, provided sample apps that illustrate and reproduce the bug, and provided a proposed patch for Glassfish 4.1 (source patch + corresponding binary) here:

[https://github.com/gregn123/GLASSFISH-20661](https://github.com/gregn123/GLASSFISH-20661)

Please see the README.txt file in the above GitHub repository for more details.
#### Environment
All
#### Affected Versions
[4.1]

### PR DESCRIPTION
**Pull request by lukasj**
Fixes #21392: Web-service endpoint is not available for the deployed EJB application

Signed-off-by: Lukas Jungmann <lukas.jungmann@oracle.com>

### CODE REVIEW

**Comment by lukasj** [comment]
@glassfishrobot Run CI tests please

**Comment by lukasj** [comment]
@glassfishrobot Run CI tests please

**Review by edburns — COMMENTED** [review]
Per discussion with Shing-wai, I will add my review in an advisory
capacity but I must point out I'm not familiar with this code and am
only able to review it from the perspective of obvious oversights and
mistakes.

I'll defer to @shingwaichan for an authoritative review.

WebServiceHandler.java:

It seems the original code did not take into account that annCtx can be
either EjbContext or EjbsContext.

Are you certain ctx.getDescriptor() will never return null, as on line
196 and others?

Are you certain name will always be a sensible value, and not null on
line 199?

A block of existing code was encapsulated in a new method getEjbName().
Are you certain the new contexts in which this method is called do not
invalidate any assumptions made by that existing code?

WebServiceProviderHandler.java

This seems to follow the same pattern as WebServiceHandler, and my same
general comments can apply.

WebServicesApplication.java

I'm not sure why removing the code was necessary, but trust it must have been.

**Review by shingwaichan — COMMENTED** [review]
The change for WebServiceHandler.java is ok. For WebServiceProviderHandler.java, it involves the extension descriptor. You may like to ask Hong to review it.

**Comment by lukasj** [comment]
wrt WebServicesApplication.java:
-there was duplicated message about deploying EJB ws endpoint and particular endpoint was processed twice later on in deployment phase

wrt WebServiceHandler.java/WebServiceProviderHandler.java

> Are you certain ctx.getDescriptor() will never return null, as on line
196 and others?

yes

> Are you certain name will always be a sensible value, and not null on
line 199?

yes, in the worst case the name will be the simple class name (the method is there to compute bean name following the EJB 3.2 spec so it can be eventually found in runtime DD)

**Comment by edburns** [comment]
Hello @lukasj.  I accept your explanations and grant my consent, but out of respect for my unfamiliarity with the code I will just leave the official review alone for others.

**Comment by arindam-bandyopadhyay** [comment]
@glassfishrobot Run CI tests please

**Comment by lukasj** [comment]
@glassfishrobot Run CI tests please

**Comment by lukasj** [comment]
@glassfishrobot Run CI tests please

File: appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServicesApplication.java [hunk 1]
```diff
@@ -155,15 +154,8 @@ private void collectEjbEndpoints(BundleDescriptor bundleDesc) {
                 }
             }
         }
-        //For ejb webservices in war we need to get the extension descriptors
```
Comment (honghzzhang) [inline appserver/webservices/jsr109-impl/src/main/java/org/glassfish/webservices/WebServicesApplication.java]: Re Shingwai's comment on extension descriptor: 
This part of the code seems for handling EJB in war case. In that case, the main bundle descriptor is WebBundleDescriptor and the EjbBundleDescriptor is an extension descriptor of the main bundle descriptor. Can you run through some test cases for EJB in war scenario where the EJB is a webservice endpoint to make sure no regressions there?

**Comment by honghzzhang** [comment]
Lucas had run additional EJB in war test cases which all passed. Approving the PR.

**Review by honghzzhang — APPROVED** [review]

**Comment by yaminikb** [comment]
@glassfishrobot run ci tests please

**Pull request merged**